src/main.cpp

```diff
@@ -4507,27 +4510,27 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
             }
             else if (inv.IsKnownType())
             {
-                CTransaction tx;
+                std::shared_ptr<const CTransaction> tx;
                 // Send stream from relay memory
                 bool push = false;
                 {
-                    LOCK(cs_mapRelay);
-                    map<uint256, CTransaction>::iterator mi = mapRelay.find(inv.hash);
+                    auto mi = mapRelay.find(inv.hash);
                     if (mi != mapRelay.end()) {
                         tx = (*mi).second;
```

Comment:
Removing `cs_mapRelay` makes the `bool push` mechanism unnecessary since the conditionals can now be chained, allowing a simpler direct `PushMessage` that would also save a `shared_ptr` copy in this case.